packages/contracts/docs/osx/02-how-to-guides/02-plugin-development/04-upgradeable-plugin/05-updating-versions.md

```diff
@@ -0,0 +1,357 @@
+      where: _payload.plugin,
+      who: _dao,
+      condition: PermissionLib.NO_CONDITION,
+      permissionId: SimpleStorageBuild2(this.implementation()).STORE_PERMISSION_ID()
+    });
+  }
+
+  /// @inheritdoc IPluginSetup
+  function implementation() external view returns (address) {
+    return simpleStorageImplementation;
+  }
+}
+```
+
+The key thing about to review in this new Plugin Setup contract is its `prepareUpdate()` function. The function only contains a condition checking from which build number the update is transitioning to build 1. Here, it is the build number 0 as this is the only update path we support. Inside, we decode the `address _account` input argument provided with `bytes _date` and pass it to the `initializeFromBuild1` function taking care of intializing the storage that was added in this build.
+
+### 3. Future builds
+
+For each build we add, we will need to add a `prepareUpdate()` function with the parameters needed to update the implementation.
+
+In this third build, for example, we are modifying the bytecode of the plugin.
+
+<details>
+<summary><code>Third plugin build example, modifying the plugin's bytecode</code></summary>
+
+```solidity
+// SPDX-License-Identifier: AGPL-3.0-or-later
+pragma solidity 0.8.17;
+
+import {IDAO, PluginUUPSUpgradeable} from '@aragon/osx/core/plugin/PluginUUPSUpgradeable.sol';
+
+/// @title SimpleStorage build 3
+contract SimpleStorageBuild3 is PluginUUPSUpgradeable {
+  bytes32 public constant STORE_NUMBER_PERMISSION_ID = keccak256('STORE_NUMBER_PERMISSION'); // changed in build 3
+  bytes32 public constant STORE_ACCOUNT_PERMISSION_ID = keccak256('STORE_ACCOUNT_PERMISSION'); // added in build 3
+
+  uint256 public number; // added in build 1
+  address public account; // added in build 2
+
+  // added in build 3
+  event NumberStored(uint256 number);
+  event AccountStored(address account);
+  error AlreadyStored();
+
+  /// @notice Initializes the plugin when build 3 is installed.
+  function initializeBuild3(
+    IDAO _dao,
+    uint256 _number,
+    address _account
+  ) external reinitializer(3) {
+    __PluginUUPSUpgradeable_init(_dao);
+    number = _number;
+    account = _account;
+
+    emit NumberStored({number: _number});
+    emit AccountStored({account: _account});
+  }
+
+  /// @notice Initializes the plugin when the update from build 2 to build 3 is applied.
+  /// @dev The initialization of `SimpleStorageBuild2` has already happened.
+  function initializeFromBuild2() external reinitializer(3) {
+    emit NumberStored({number: number});
+    emit AccountStored({account: account});
+  }
+
+  /// @notice Initializes the plugin when the update from build 1 to build 3 is applied.
+  /// @dev The initialization of `SimpleStorageBuild1` has already happened.
+  function initializeFromBuild1(address _account) external reinitializer(3) {
+    account = _account;
+
+    emit NumberStored({number: number});
+    emit AccountStored({account: _account});
```

Comment:
Combine in just one function for general initialization and one function for upgrades:
https://github.com/aragon/simple-storage-example-plugin/blob/main/contracts/release1/build3/SimpleStorageR1B3.sol#L30-L50

Reply:
Simple storage plugin update - added to notes!